### PR TITLE
Backfill job for section instructors (TEACH-682)

### PR DESCRIPTION
**Pull request by etaderhold**
This is a oneoff batch job to backfill the section instructors table with one row for the owner of each section. 

It uses the `Section.ensure_owner_is_active_instructor` function that has been used as an `after_save` hook for new and updated sections in production for the past week. This function checks to see whether there is already an active SectionInstructor object for the section owner and creates one if not.

During testing for this backfill job I noticed that the `ensure_owner_is_active_instructor` method was querying the database to pull in the section and user objects all over again after saving, as part of the validation. I change that in this PR to do an `insert` instead of `create` to skip this validation step and save two queries per section processed.

## Links

- [JIRA](https://codedotorg.atlassian.net/browse/TEACH-682)

## Testing story

The meat of this script is already tested through the existing tests on `ensure_owner_is_active_instructor`. The backfill job is basically a wrapper around that.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/sections/section.rb [hunk 1]
```diff
@@ -234,7 +234,17 @@ def ensure_owner_is_active_instructor
 
     si = SectionInstructor.with_deleted.find_by(instructor: user, section_id: id)
     if si.blank?
-      SectionInstructor.create!(section_id: id, instructor: user, status: :active)
+      # Using insert instead of create saves on queries to re-read the section
+      # and user from the DB to validate these are still actual objects.
```
Comment (cat5inthecradle) [inline dashboard/app/models/sections/section.rb:240]: Is this the ideal long-term implementation? Are there other consequences of using the lower-level `insert`? Totally makes sense for the backfill task.

Comment (etaderhold) [reply dashboard/app/models/sections/section.rb:240]: Long-term, after the backfill, this insert _should_ only be running when a new section is created. The only way I could see `insert` going badly compared to `create` is if the user somehow managed to delete their account in the fraction of a second between hitting the "create section" button and that section being saved to the database.

Comment (sureshc) [reply dashboard/app/models/sections/section.rb:240]: Is the performance optimization worth the absence of callbacks and validations running here?

Comment (etaderhold) [reply dashboard/app/models/sections/section.rb:240]: > Is the performance optimization worth the absence of callbacks and validations running here?

What do you think?

The validations query the database to make sure that the `Section` and `User` we just queried less than a second ago still exist, before we can write a `SectionInstructor` that references them. Skipping those validations will save about 8 million queries during this backfill, and the downside is that we risk storing a dangling `SectionInstructor` row in the database for every user who happens to delete their account at the exact instant this backfill job tries to fix up one of their sections. What do you think the right trade-off is here?

Comment (bethanyaconnor) [reply dashboard/app/models/sections/section.rb:240]: The other, more long-term downside I see is someone expecting callbacks to be run and then them not being run (I may or may not have been burned by this in the past...). That said, `SectionInstructor` is a pretty small model and I'm having trouble thinking of too many validations/callbacks that would need to be added.

For the backfill, this changes makes sense to me

Comment (etaderhold) [reply dashboard/app/models/sections/section.rb:240]: I see your point about this perhaps causing a hook-related bug down the line. I'll plan to switch it back after running the backfill.

Any other concerns blocking this PR being merged?

Comment (bethanyaconnor) [reply dashboard/app/models/sections/section.rb:240]: not from me!

**Review by bethanyaconnor — APPROVED** [review]

**Pull request merged**